src/runtime/crt/memory/stack_allocator.c

```diff
@@ -16,17 +16,22 @@
  * specific language governing permissions and limitations
  * under the License.
  */
-
 // LINT_C_FILE
-
 #include <tvm/runtime/crt/stack_allocator.h>
+#ifdef TVM_CRT_DEBUG
+#include <tvm/runtime/crt/logging.h>
+#endif
 
 void* StackMemoryManager_Allocate(tvm_workspace_t* tvm_runtime_workspace, int32_t nbytes) {
   uint32_t offset_bytes = (~nbytes + 1) & (TVM_RUNTIME_ALLOC_ALIGNMENT_BYTES - 1);
   uint8_t* current_alloc = tvm_runtime_workspace->next_alloc;
   uint8_t* next_alloc = tvm_runtime_workspace->next_alloc + nbytes + offset_bytes;
   uint8_t* workspace_end = tvm_runtime_workspace->workspace + tvm_runtime_workspace->workspace_size;
-
+#ifdef TVM_CRT_DEBUG
```

Comment:
I can see why this is not strictly production code, but I don't think that this should be on only in debug mode. it's pretty easy to get memory corruption in a production system, so i'd propose this is just always-on or disable-able by `#define`.

Reply:
I disagree -- user does not write the code that uses this -- its generated by tvm -- if tvm generates code that does not follow the pattern it should be fixed in the codegenerator -- the tree-scoped tir.allocates always produce the code in this manner. Therefore, I think the 4-byte overhead for each allocate should only be enabled to debug the codegenerator -- which I don't think a user would be doing -- therefore default should be off, IMO.

This is aimed to be used in a deployment scenario.

Reply:
i would push back with: in a deployment scenario, what `$pc` do you want logged when there is memory corruption?

Reply:
I agree if illegal code is "generated", it will result in garbage calculation or exceeding the provided buffer for allocator if we are using this to handle workspace data.

Even this gives an error (of illegal free) that should mean a bug in the codegen (a very unlikely one if everything is lowered from TIR). In which case, the codegen developer needs to fix. Therefore, I do think the developer should investigate with DEBUG mode to see if this is causing the error and fix the codegen.

Reply:
So there are two main issues here:
* We are using 4 bytes per block to check memory. Given the memory constraints on some devices, it might be too expensive. 
* More importantly, we are loosing the block alignment. There are some microcontrollers that would not allow non-aligned memory access. 
For those two reasons, I would disable it by default and only enable for specific reasons. I can use a different MACRO (e.g., STACK_ALLOCATOR_CHECK_ENABLED) that people can use to enable this feature, so it would not be only for debug. What do you think? Also adding @manupa-arm and @Mousius to the discussion

Reply:
the case i'm thinking about is where other code running on the user device overwrites e.g. `next_alloc` or the `ptr` passed to Free(). in this case, the error is likely in the user application rather than in TVM codegen, and this allocator would be a bit of a black box (of course, they could inspect the source code, but it would be helpful to provide an indication of memory corruption at the earliest point it could be detected).

Reply:
I'm not following what you mean by "other code" ? The intention of this to be used by TVMBAW calls generated by TVM codegen only. tvm_runtime_workspace->next_alloc should only be accessed/written by StackMemoryManager_Allocate and Free. 

Are you talking about a scenario where tvm_runtime_workspace is accidentally accessed by other user code ?. (Im trying to figure out a scenario how this error is triggered outside of running generated code)

Reply:
for example, perhaps you are integrating this with an application which operates a radio, but _that_ library accidentally overwrites `next_alloc`. in this case, it's preferable to have our library fail predictably as soon as it detects memory corruption. without the check here, we would hand out a corrupt memory pointer on the following allocate call.

anyway--sorry, my comment here was mainly aimed at addressing @giuseros question about unit testing. this my rationale for solving the problem of how to unit test this by making it on-by-default. we should probably address productionization in a different PR, so to move forward here, we can just find another way to enable the check in unit tests--that would address my primary concern here, which is to prove that the generated code uses the APIs correctly.

Reply:
I get the negative consequences (therefore agree that we need a check) but not sure how realistic are they to be on by default. If we really want to have them by default -- @giuseros I'd suggest we'd need a seperate buffer to hold debug info not interleaved with the data buffer.

We can always compile with -D STACK_ALLOCATOR_CHECK_ENABLED to unit test, is there a concern of not being able to unit test because of its not on by default.

Reply:
> this example is accessing the next_alloc via some pointer arithmetic on a nearby data that got linked in in the .data section?

yeah exactly. common cause is passing in a bad pointer to some other data structure which then causes a library to corrupt the data. on x86, e.g. dereferencing a nullptr (e.g. trying to access a struct member with struct ptr==0) causes a segfault; on bare metal, you just read whatever program data happens to be there (often .text or from the SoC ROM).

one thing with the current impl is that currently `tag` is just placed in the extra space given by `TVM_CRT_ALIGNMENT_BYTES`, so because that's 16, there is always an extra word of memory. the overhead here is a few cycles of compute. I don't see it as particularly large considering we are doing a lot more compute when e.g. doing a conv2d.

> is there a concern of not being able to unit test because of its not on by default ?
yeah exactly--but it would be fine to just turn it on for unit tests too. If you want to do that, perhaps we should create a `crt_config.h` for unit tests and then change the include path used for unit tests in StandaloneCrt.cmake and in the python code which builds the AOT makefile.

Reply:
I am not following the last comment. If you store X bytes of memory, you allocate a block of [X/16]*16 bytes. Then you *additionally* allocate 4 bytes for the tag. So I think there are two issues:

* Additional 4 bytes of memory per block allocated
* Loss of memory alignment. This can really hurt accelerators that usually have strong alignment requirements. 

For these reasons I would leave it off by default, and only enable it for debugging. If we want to add a feature to enforce the FIFO ordering  in production I would leave it  for a subsequent and more focused PR 

About the tests, I  tried to follow your suggestion to have a local `crt_config.h`, but the `stack_allocator.c` is compiled in `libmemory.a` that includes the default `crt_config.h`.

Reply:
ah i'm sorry--my previous comment was incorrect. you're indeed right that this checking mode wastes 4 bytes per tensor. I understand that we especially need to be cognizant of the alignment issue, so how about this proposal then:

- the libmemory.a used in `make crttest` is built with the `crt_config.h` used to build `standalone_crt` for:
   1. the host machine
   2. tests launched from `test_crt.py`
- in these cases, we can afford the extra cycles for a check, and I think this shouldn't affect test results since we aren't using accelerators from here.
- so my proposal is: let's include this check by defining a flag in `src/runtime/crt/host/crt_config.h`. we can also include a commented copy of this flag in the template `crt_config.h`, but leave it off there. finally, i'm not sure we should use `TVM_CRT_DEBUG`, since that may be spammy in unit tests; but perhaps we could define another flag?

let me know how this sounds to you. the main thing i'd like to achieve is that unit tests of the CRT and of the AOT that are executed in CI are run with this check enabled. 

btw, we should probably discuss moving the alignment rules to a higher level than the runtime memory allocator. I think it's fine to leave it there for this PR, but seems like it should be part of our higher-level memory planning work.

Reply:
Cool! I agree that alignment should be considered in a higher-level memory planning work. :). 
However, as shown here, its trivial to support a global alignment (that addressed produced by the allocator is aligned to a global granularity) might be beneficial in the mean time.

Reply:
Yes, I agree on the approach! I just pushed the new tests (with the check enabled). The name of the flag I picked is `TVM_CRT_STACK_ALLOCATOR_ENABLE_FIFO_CHECK`, please let me know if you agree with it, or if you have a better name (naming is not my strong skill :) )